### DISCUSSION TITLE
Host HID example how to control caps lock LED on USB keyboard

### DISCUSSION THREAD

**Discussion by jeroentaverne** [hathach/tinyusb#1191] [General]
Using the host_cdc_msc_hid example I would also like to control the capslock LED on a USB keyboard. What lines of code would I need to add to control it? Turning it on would already be great.

**Comment by hathach** [comment]
you can use the following API. typically for boot mode, `report_id = 0, report_type = HID_REPORT_TYPE_OUTPUT, report = led bitmask (1 for each LED), len = 1`

```C
// Set Report using control endpoint
// report_type is either Intput, Output or Feature, (value from hid_report_type_t)
bool tuh_hid_set_report(uint8_t dev_addr, uint8_t instance, uint8_t report_id, uint8_t report_type, void* report, uint16_t len);

typedef enum
{
  KEYBOARD_LED_NUMLOCK    = TU_BIT(0), ///< Num Lock LED
  KEYBOARD_LED_CAPSLOCK   = TU_BIT(1), ///< Caps Lock LED
  KEYBOARD_LED_SCROLLLOCK = TU_BIT(2), ///< Scroll Lock LED
  KEYBOARD_LED_COMPOSE    = TU_BIT(3), ///< Composition Mode
  KEYBOARD_LED_KANA       = TU_BIT(4) ///< Kana mode
}hid_keyboard_led_bm_t;
```

**Reply by borb** [reply]
Hello! I tried this:

```c
leds = KEYBOARD_LED_CAPSLOCK;
tuh_hid_set_report(dev_addr, instance, 0, HID_REPORT_TYPE_OUTPUT, &leds, sizeof(leds));
```

But I'm not seeing anything happen on the keyboard (it's definitely working elsewhere in the code). I've tried report types of both OUTPUT and FEATURE (on the off chance), but no such luck with the LED status.

(leds is `uint8_t`)

**Reply by dquadros** [reply]
This was not working for me until I realized that 'leds' was a local variable in my code and would be overrun when the routine it was in returned. I changed it to 'static' and it worked as a charm. 

Not sure if I should call tuh_hid_set_report inside the report received callback, but it is working. Testing done in a Pi Pico.

**Comment by jeroentaverne** [comment]
Thanks, I will give it a try.

**Comment by No0ne** [comment]
@hathach Could you please provide a full example so I can fix my code [here](https://github.com/No0ne/ps2pico/blob/main/ps2pico.c#L130) and [here](https://github.com/No0ne/ps2x2pico/blob/main/ps2x2pico.c#L156), thanks!

**Reply by dquadros** [reply]
While this is not a didactic example, and only uses the CAPS LOCK LED, it does show how to control the LEDs:

https://github.com/dquadros/RPTerm/blob/main/keybd.cpp
 
It also shows a way to implement auto-repeat.

**Reply by No0ne** [reply]
Thanks!
